### DISCUSSION TITLE
How to enable follow-mode for helm-grep-do-git-grep

### DISCUSSION THREAD

**Discussion by zw963** [emacs-helm/helm#2524] [General]
Hi, i want to use helm builtin `helm-grep-do-git-grep` instead of not live maintained helm plugins `helm-git-grep-at-point`

but, i don't know how to enable follow-mode for this command  in my emacs config.

For now, i have to C-c C-f switch follow-mode after i restart my emacs.

Thank you.

**Comment by thierryvolpiatto** [comment]
"Billy.Zheng" ***@***.***> writes:

> Hi, i want to use helm builtin helm-grep-do-git-grep instead of not live maintained helm plugins helm-git-grep-at-point
>
> but, i don't know how to enable follow-mode for this command in my emacs config.
>
> For now, i have to C-c C-f switch follow-mode after i restart my emacs.

For all helm commands, not just *git-grep, add to your init file:

    (setq helm-follow-mode-persistent t)

Then once you hit once C-c C-f in a specific source, you will have
follow-mode enabled for this source forever until you hit again C-c C-f.


-- 
Thierry

**Reply by zw963** [reply]
> (setq helm-follow-mode-persistent t)

Yes, in fact, i set this to true, but, i still have to C-c C-f once since i first start emacs, because i don't know how to enable
following mode to `helm-grep-do-git-grep` directly in my config.

What i expected to is, enable follow-mode for helm-grep-do-git-grep after i start emacs immediately.

I guess this source is name `helm-source-grep-git`, but, i found it is nil when i try to retrieve this variable.

**Comment by zw963** [comment]
I try to enable it with:

```cl
(add-hook 'helm-before-initialize-hook
          (lambda () (helm-attrset 'follow t helm-source-grep-git)))
```

But this config even broken my helm config ...

**Comment by thierryvolpiatto** [comment]
"Billy.Zheng" ***@***.***> writes:

>     (setq helm-follow-mode-persistent t)
>    
> Yes, in fact, i set this to true, but, i still have to C-c C-f once since i first start emacs, because i don't know how to enable
> following mode to helm-grep-do-git-grep directly in my config.

You don't have to do this, please follow these steps:

    (setq helm-follow-mode-persistent t) ;; Eval it with C-x C-e.

If you don't eval it it will take effect only in next emacs session (it
is what is happening to you right now).

Then launch helm-grep-do-git-grep somewhere and hit C-c C-f, that's all
you don't have to do anything else.

Restart emacs, launch helm-grep-do-git-grep somewhere you will see it
have follow-mode enabled.


-- 
Thierry

**Comment by thierryvolpiatto** [comment]
"Billy.Zheng" ***@***.***> writes:

> I try to enable it with:
>
> (add-hook 'helm-before-initialize-hook
>           (lambda () (helm-attrset 'follow t helm-source-grep-git)))

NO!

> But this config even broken my helm config ...

Of course.


-- 
Thierry

**Comment by thierryvolpiatto** [comment]
Also, note that once follow-mode is enabled for your source, if you hit another time C-c C-f it will disable follow-mode but persistently, it will be disabled next time you start emacs.

**Comment by zw963** [comment]
> You don't have to do this, please follow these steps:

>    (setq helm-follow-mode-persistent t) ;; Eval it with C-x C-e.

> If you don't eval it it will take effect only in next emacs session (it
> is what is happening to you right now).

> Then launch helm-grep-do-git-grep somewhere and hit C-c C-f, that's all
> you don't have to do anything else.

> Restart emacs, launch helm-grep-do-git-grep somewhere you will see it
> have follow-mode enabled.

Yes, i find it works! thank you.

Though, i am curious, how helm can save my status cross emacs session.

ASAK, after i reboot emacs, all status should missing.

Does helm save status to a file in ~/.emacs.d/ folder?

**Comment by thierryvolpiatto** [comment]
"Billy.Zheng" ***@***.***> writes:

> Yes, i find it works! thank you.

Great.

> Though, i am curious, how helm can save my status cross emacs session.
>
> ASAK, after i reboot emacs, all status should missing.
>
> Does helm save status to a file in ~/.emacs.d/ folder?

In your custom file.

-- 
Thierry
